pkg/ruler/rules/local/local.go

```diff
@@ -0,0 +1,136 @@
+package local
+
+import (
+	"context"
+	"flag"
+	"io/ioutil"
+	"path"
+
+	"github.com/pkg/errors"
+
+	rulefmt "github.com/cortexproject/cortex/pkg/ruler/legacy_rulefmt"
+	"github.com/cortexproject/cortex/pkg/ruler/rules"
+)
+
+type Config struct {
+	Directory string `yaml:"directory"`
+}
+
+// RegisterFlags registers flags.
+func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.StringVar(&cfg.Directory, prefix+"local.directory", "", "Directory to scan for rules")
+}
+
+// Client expects to load already existing rules located at:
+//  cfg.Directory / userID / namespace
+type Client struct {
+	cfg Config
+}
+
+func NewLocalRulesClient(cfg Config) (*Client, error) {
+	if cfg.Directory == "" {
+		return nil, errors.New("directory required for local rules config")
+	}
+
+	return &Client{
+		cfg: cfg,
+	}, nil
+}
+
+// ListAllRuleGroups implements RuleStore
+func (l *Client) ListAllRuleGroups(ctx context.Context) (map[string]rules.RuleGroupList, error) {
+	lists := make(map[string]rules.RuleGroupList)
+
+	root := l.cfg.Directory
+	infos, err := ioutil.ReadDir(root)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to read dir "+root)
+	}
+
+	for _, info := range infos {
+		if !info.IsDir() {
+			continue
+		}
+
+		list, err := l.listAllRulesGroupsForUser(ctx, info.Name())
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to list rule group")
+		}
+
+		lists[info.Name()] = list
+	}
+
+	return lists, nil
+}
+
+// ListRuleGroups implements RuleStore
+func (l *Client) ListRuleGroups(ctx context.Context, userID string, namespace string) (rules.RuleGroupList, error) {
+	if namespace != "" {
+		return l.listAllRulesGroupsForUserAndNamespace(ctx, userID, namespace)
+	}
+
+	return l.listAllRulesGroupsForUser(ctx, userID)
+}
+
+// GetRuleGroup implements RuleStore
+func (l *Client) GetRuleGroup(ctx context.Context, userID, namespace, group string) (*rules.RuleGroupDesc, error) {
+	return nil, errors.New("GetRuleGroup unsupported in rule local store")
+}
+
+// SetRuleGroup implements RuleStore
+func (l *Client) SetRuleGroup(ctx context.Context, userID, namespace string, group *rules.RuleGroupDesc) error {
+	return errors.New("SetRuleGroup unsupported in rule local store")
+}
+
+// DeleteRuleGroup implements RuleStore
+func (l *Client) DeleteRuleGroup(ctx context.Context, userID, namespace string, group string) error {
+	return errors.New("DeleteRuleGroup unsupported in rule local store")
+}
+
+func (l *Client) listAllRulesGroupsForUser(ctx context.Context, userID string) (rules.RuleGroupList, error) {
+	var allLists rules.RuleGroupList
+
+	root := path.Join(l.cfg.Directory, userID)
```

Comment:
You should use `filepath.Join()` instead of `path.Join()` when building filesystem paths.

Reply:
Thank you for your patient reviews.  These mistakes feel silly to me.  I have a lot to learn about Go.